gravitee-am-factor/gravitee-am-factor-recovery-code/src/main/java/io/gravitee/am/factor/recovery/code/provider/RecoveryCodeFactorProvider.java

```diff
@@ -0,0 +1,131 @@
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.gravitee.am.factor.recovery.code.provider;
+
+import io.gravitee.am.common.exception.mfa.InvalidCodeException;
+import io.gravitee.am.common.utils.SecureRandomString;
+import io.gravitee.am.factor.api.Enrollment;
+import io.gravitee.am.factor.api.FactorContext;
+import io.gravitee.am.factor.api.FactorProvider;
+import io.gravitee.am.factor.api.RecoveryFactor;
+import io.gravitee.am.factor.recovery.code.RecoveryCodeFactorConfiguration;
+import io.gravitee.am.gateway.handler.root.service.user.UserService;
+import io.gravitee.am.identityprovider.api.DefaultUser;
+import io.gravitee.am.model.Factor;
+import io.gravitee.am.model.factor.EnrolledFactor;
+import io.gravitee.am.model.factor.EnrolledFactorSecurity;
+import io.gravitee.am.model.factor.FactorStatus;
+import io.reactivex.Completable;
+import io.reactivex.Single;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+
+import static io.gravitee.am.common.factor.FactorSecurityType.RECOVERY_CODE;
+
+/**
+ * @author Ashraful Hasan (ashraful.hasan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class RecoveryCodeFactorProvider implements FactorProvider, RecoveryFactor {
+    private static final Logger logger = LoggerFactory.getLogger(RecoveryCodeFactorProvider.class);
+
+    @Autowired
+    private RecoveryCodeFactorConfiguration configuration;
+
+    @Override
+    public Completable sendChallenge(FactorContext context) {
+        return Completable.complete();
+    }
+
+    @Override
+    public Completable verify(FactorContext context) {
+        final String code = context.getData(FactorContext.KEY_CODE, String.class);
+        final EnrolledFactor enrolledFactor = context.getData(FactorContext.KEY_ENROLLED_FACTOR, EnrolledFactor.class);
+        final List<String> recoveryCodes = (List<String>) enrolledFactor.getSecurity().getAdditionalData().get(RECOVERY_CODE);
+
+        return Completable.create(emitter -> {
+            if (recoveryCodes.contains(code)) {
+                //remove the code from the list as the recovery is not re-usable
+                recoveryCodes.remove(code);
+                enrolledFactor.getSecurity().setAdditionalData(Map.of(RECOVERY_CODE, recoveryCodes));
+                emitter.onComplete();
+            } else {
+                logger.error("An error occurred while validating with recovery code");
```

Comment:
Don't think it is something we should log as ERROR. An invalid recovery code is a standard case.
IMHO, it is at most a WARN but INFO or DEBUG may be fine.

Reply:
error log removed.